### PR TITLE
[KED-3004] Search MVP experiment tracking

### PR DESCRIPTION
**Pull request by studioswong**
## Description

resolves [KED-3004](https://jira.quantumblack.com/browse/KED-3004).

This PR is the work in setting up the search for experiment tracking runslist. This will now allow users to search through the runs list and return results with runs that either the 'Title', 'gitsha' and 'notes' field matches the search value.

image

image

## Development notes

1. I have refactored the existing search bar from a tailored component to the node-list to a general search bar, and set up related tests.
2. I have refactored existing runs-list and runs-list card setup with the search Logic, and added in tests related to the search feature.
3. I have set up a new test suite for runs-list component as currently there isn't any test suites for that component. 

## QA notes

Please check out the branch and QA the search. I will also sync up with design (@gabrielcomym) to go through the latest implementation. 

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [x] Added new entries to the `RELEASE.md` file
- [x] Added tests to cover my changes

### CODE REVIEW

File: src/components/experiment-tracking/runs-list/runs-list.scss
Comment (tynandebold) [inline src/components/experiment-tracking/runs-list/runs-list.scss:28]: These exact styles are also present in the `/sidebar/sidebar.scss` file. Not sure why they're duplicated here?

image

File: src/components/experiment-tracking/runs-list/runs-list.test.js
Comment (tynandebold) [inline src/components/experiment-tracking/runs-list/runs-list.test.js:24]: Maybe we should also test when a bookmark is `true` and then we get the two-accordions view?

File: src/components/search-list/search-list.scss
Comment (tynandebold) [inline src/components/search-list/search-list.scss:6]: The formatting is off on this file. Can you try and save it again?

File: src/utils/index.js
Comment (tynandebold) [inline src/utils/index.js:66]: `search value` is two words and you don't need the `, true if no search value given` in the top line since you specify that in the return line.

```suggestion
 * Check whether a piece of text matches the search value
```

**Review by tynandebold — CHANGES_REQUESTED** [review]
Awesome stuff. Excellent first pass.

A few comments from me, but nothing major. They should be quick changes :)

File: src/components/experiment-tracking/runs-list-search/runs-list-search.js
Comment (studioswong) [reply src/components/experiment-tracking/runs-list-search/runs-list-search.js:11]: good spot - I completely forgot about it

File: src/components/experiment-tracking/runs-list-card/runs-list-card.js
Comment (studioswong) [reply src/components/experiment-tracking/runs-list-card/runs-list-card.js:38]: thanks for suggesting this! Actually I originally had this set up exactly the way you suggested but then striped this out into the 2 separate checks as I find separating the logic into 2 separate functions more readable ( and also as a result of comments from previous PRs 😆  it's funny how many difference of opinion there is). 

This is the same behind setting out a separate variable for the note check for simplicity and readability in the JSX, without the need for anyone later in to refer to what `isMatchSearchValue` does as well.

File: src/components/experiment-tracking/runs-list/runs-list.scss
Comment (studioswong) [reply src/components/experiment-tracking/runs-list/runs-list.scss:28]: good catch, I was about to delete the style under `sideBar` but forgot!

**Review by rashidakanchwala — COMMENTED** [review]
So I tried the search and it works well. But I noticed the search results number doesn't change. Please see it shows Bookmarked (2) Unbookmarked(5) --> instead of Bookmarked (1) Unbookmarked(1)

Screenshot 2022-01-21 at 09 44 47

**Comment by tynandebold** [comment]
> So I tried the search and it works well. But I noticed the search results number doesn't change. Please see it shows Bookmarked (2) Unbookmarked(5) --> instead of Bookmarked (1) Unbookmarked(1)
> 
> Screenshot 2022-01-21 at 09 44 47

Great catch.

**Comment by studioswong** [comment]
good catch @rashidakanchwala - I'll update this

File: src/components/experiment-tracking/runs-list/runs-list.test.js
Comment (studioswong) [reply src/components/experiment-tracking/runs-list/runs-list.test.js:24]: indeed - on that note, I have observed that there are currently no tests set up for the run-list for the bookmark journey - it would be great if we can set some tests for that. I will raise a separate ticket for that. 

Your suggested test condition would sound like a perfect test case for the bookmark journey test setup to be dealt with in that ticket - what do you think?

Comment (tynandebold) [reply src/components/experiment-tracking/runs-list/runs-list.test.js:24]: Yeah we could and probably should be robust and create a separate ticket tailored more towards bookmarks. We'll still need a test for the distinct accordions, and I think that could live in this file, as those are part of the runs list.

You can do it so quick too, because if you use `bookmark: true` on one of the runs then the condition on line 38 becomes this, and then that basically checks for two accordions:

```js
expect(wrapper.find('.runs-list__wrapper').length).toBe(2);
```

File: src/components/experiment-tracking/runs-list/runs-list.scss
Comment (tynandebold) [inline src/components/experiment-tracking/runs-list/runs-list.scss:30]: I think this should be `padding: 3.5em 3.5em 1.2em 3em;`. `35px` (plus the little extra from the search component) between the search and compare toggle, not `20px`.

File: src/components/experiment-tracking/runs-list-card/runs-list-card.js
Comment (tynandebold) [reply src/components/experiment-tracking/runs-list-card/runs-list-card.js:38]: Fine with me. I would still change the variable name to `isSearchValueInNotes` instead of `isNotesContainSearchValue`.

**Review by tynandebold — APPROVED** [review]
Looks awesome. I'm approving with a couple of comments that I hope you'll change before merging in (specifically the spacing between the search and the comparison toggle and the `All` label never appearing).

Great stuff!

**Review by rashidakanchwala — APPROVED** [review]
Thanks, Susanna! Looks great.

**Pull request merged**
